### DISCUSSION TITLE
Failing at managing overlapping patterns with priority

### DISCUSSION THREAD

**Discussion by rpdelaney** [Woundorf/foxreplace#367] [Q&A] [answered]
Here's a substitution group:

<details><summary>FoxReplace.json</summary>

```json
{
  "version": "2.6",
  "groups": [
    {
      "name": "NFT",
      "urls": [],
      "substitutions": [
        {
          "input": "\\(non-fungible token(s)?\\)",
          "inputType": "regexp",
          "outputType": "text",
          "output": "(yes, you read that correctly)",
          "caseSensitive": false
        },
        {
          "input": "\\(NFT(s)?\\)",
          "inputType": "regexp",
          "outputType": "text",
          "output": "(yes, you read that correctly)",
          "caseSensitive": false
        },
        {
          "input": "non-fungible tokens",
          "inputType": "wholewords",
          "outputType": "text",
          "output": "URLs on blockchain",
          "caseSensitive": false
        },
        {
          "input": "non-fungible token",
          "inputType": "wholewords",
          "outputType": "text",
          "output": "URL on blockchain",
          "caseSensitive": false
        },
        {
          "input": "NFTs",
          "inputType": "wholewords",
          "outputType": "text",
          "output": "URLs on blockchain",
          "caseSensitive": false
        },
        {
          "input": "an NFT",
          "inputType": "wholewords",
          "outputType": "text",
          "output": "a URL on blockchain",
          "caseSensitive": true
        },
        {
          "input": "An NFT",
          "inputType": "wholewords",
          "outputType": "text",
          "output": "A URL on blockchain",
          "caseSensitive": false
        },
        {
          "input": "NFT",
          "inputType": "wholewords",
          "outputType": "text",
          "output": "URL on blockchain",
          "caseSensitive": false
        }
      ],
      "html": "none",
      "enabled": true,
      "mode": "auto&manual"
    }
  ]
}
```

</details>

With this activated I mosey on over to [NFT on Wikipedia](https://en.wikipedia.org/wiki/Non-fungible_token).

Before applying substitutions:

> A non-fungible token (NFT) is a unique digital identifier that is recorded on a blockchain and is used to certify ownership and authenticity.

After applying substitutions:

> A URL on blockchain (URL on blockchain) is a unique digital identifier that is recorded on a blockchain and is used to certify ownership and authenticity.

That's not what I expected, though. The pattern with second-highest priority is `\(NFT(s)?\)` which should match first, and thus be replaced with `(yes, you read that correctly)` before the final pattern is reached.

Let's compare to the [Investopedia article on NFT](https://www.investopedia.com/non-fungible-tokens-nft-5115211).

Before applying substitutions:

> What Is a Non-Fungible Token (NFT)? 
> Non-fungible tokens (NFTs) are assets that have been tokenized via a blockchain. 

After applying substitutions:

> What Is a URL on blockchain (yes, you read that correctly)?
> URLs on blockchain (yes, you read that correctly) are assets that have been tokenized via a blockchain.

That _is_ what I expected. So what's different with Wikipedia? 🤔

**Comment by Woundorf** [comment, marked as answer]
Because in Wikipedia `NFT` is inside `<b>` tags, but the parentheses are outside:

![imatge](https://github.com/Woundorf/foxreplace/assets/1915151/f777b15c-085e-4763-a7fe-5628d6b45cd9)

Since substitutions are applied element by element, your second regex does not match.

**Reply by rpdelaney** [reply]
Arg. Am I stuck parsing HTML with regular expressions? Can you recommend another way by any chance?

Thank you for the explanation regardless. I love this plugin -- it makes the internet so much more fun

**Reply by Woundorf** [reply]
Unfortunately, at the moment, yes, parsing HTML with regular expressions is the only option. And at the cost of some pages making heavy use of JavaScript probably breaking at least partially (activating the "replace scripts" options improves this but not fully).

There is a suggestion (#211) to replace text over multiple nodes that would solve this case and allow your substitutions to work as they are now, but it is not implemented.
